### ISSUE TITLE
mpl_connect does not write output to jupyter notebook

### ISSUE THREAD

**Issue by michalsustr** [matplotlib/matplotlib#6948]
Hi, I have:
- notebook 4.2.1
- matplotlib 1.5.1
- Python 3.5.2
- running on Ubuntu

I setup my env through conda.

An example from the docs that uses mpl_connect: http://matplotlib.org/users/event_handling.html#event-attributes

```
from matplotlib import pyplot as plt

class LineBuilder:
    def __init__(self, line):
        self.line = line
        self.xs = list(line.get_xdata())
        self.ys = list(line.get_ydata())
        self.cid = line.figure.canvas.mpl_connect('button_press_event', self)

    def __call__(self, event):
        print('click', event)
        if event.inaxes!=self.line.axes: return
        self.xs.append(event.xdata)
        self.ys.append(event.ydata)
        self.line.set_data(self.xs, self.ys)
        self.line.figure.canvas.draw()

fig = plt.figure()
ax = fig.add_subplot(111)
ax.set_title('click to build line segments')
line, = ax.plot([0], [0])  # empty line
linebuilder = LineBuilder(line)

plt.show()
```

The `print('click', event)` line does not show any output in my Jupyter notebooks.

**Comment by WeatherGod** [comment]
Eh, I think it is a bit more complicated than that. I bet if you clicked
enough times, it would show up due to buffered outputs, but I don't know
that for certain. It would help to know what backend you are using for this
example.

**Comment by michalsustr** [comment]
Repeated clicks don't help. What do you mean by backend? Please specify what you are interested in (packages/versions/ dunno). I recently started using matplotlib/jupyter so I'm not so well versed in jargon :)

**Comment by WeatherGod** [comment]
what is returned by the following cell:

import matplotlib
matplotlib.get_backend()

As for repeated clicks, I am betting it would have to be on the order of
about 100 or so clicks to trigger a buffer clear, so unless you clicked
that many times, I still am betting that is the problem.

**Comment by michalsustr** [comment]
it uses 'nbAgg'

> I am betting it would have to be on the order of about 100 or so clicks to trigger a buffer clear

That's what I would call a bug :) I'm quite confident it's over 100. https://imgur.com/a/bzCSJ

**Comment by tacaswell** [comment]
The problem is that nbagg does not hook into the widget framework (as it existed first :wink: ) and the stdout from the callbacks are not getting set across the com.  Pretty sure this is fixed by `ipympl`  https://github.com/matplotlib/jupyter-matplotlib

**Comment by tacaswell** [comment]
You should also use `draw_idle` instead of `draw`

**Comment by michalsustr** [comment]
Okay, so I pip installed 
`pip install ipympl`
and
`jupyter nbextension enable --py --sys-prefix ipympl`
from the link. When I relaunched the notebook without my custom init, using matplotlib backend: TkAgg, it opens a new window and outputs click events into it. But if I specify magic
`%matplotlib notebook`
it is still the same, no change. And I added the `draw_idle`, that's nice.

**Comment by michalsustr** [comment]
There seems to be another issue: I have two notebooks opened now: A and B.

In A, I have %matplotlib tk 
In B, I have %matplotlib notebook

If I try to change A: %matplotlib notebook I get
> Warning: Cannot change to a different GUI toolkit: notebook. Using tk instead.
If I try to change B: %matplotlib tk I get
> Warning: Cannot change to a different GUI toolkit: tk. Using notebook instead.

Changing to `inline` works fine. I thought these can be changed dynamically?

**Comment by jenshnielsen** [comment]
Matplotlib has never supported changing the backend in an running session. You will need to restart the notebook to change backend. 

I don't think `%matplotlib notebook` loads the new `ipympl` backend yet. You will need to use it as described in https://github.com/matplotlib/jupyter-matplotlib/blob/master/examples/ipympl.ipynb that should probably be fixed but I think the fix needs to go into IPython where the % magics are defined

**Comment by WeatherGod** [comment]
No, interactive backends can not always be changed (we are working to make
that more possible, though). Once pyplot is imported, you can't switch
interactive backends. Our eventual goal is that switching would be possible
until the first figure creation, and ideally that it could be deferred
until the first show() (but that won't be possible for a long time).

**Comment by michalsustr** [comment]
Thanks. After about an hour of shuffling around imports, restarting kernels etc. I was able to come up with this:

![image](https://cloud.githubusercontent.com/assets/1756494/17706211/71ac9ac0-63db-11e6-86e4-8f71311b6ca9.png)

Also, the correct order for the imports is:
```
# standard imports
import ipympl
import matplotlib.pyplot as plt
import matplotlib
from pylab import *
# other bunch of things I use
import math
import random
from tqdm import tqdm
import logging
import numpy as np
import pandas as pd
```

So, currently the backend is: 
```Using matplotlib backend: TkAgg```

However as you can see on the uploded img, it doesn't integrate that well with jupyter ntb... but I get the output there, and it also has export button, which is a nice feature.

I think it would be better to have this somewhere documented, because I wonder how I would come up with this without submitting the issue. I'd like to add it somewhere, but I'm not sure where, since I don't know your ecosystem that well...

**Comment by michalsustr** [comment]
And I almost forgot, there was that unhandled UserWarning...

**Comment by WeatherGod** [comment]
I *strongly* urge against the usage of pylab. At the very least, it doesn't
make sense to mix pylab and pyplot. Use one or the other -- I recommend
pyplot because "star imports" are risky (and did you know that it messes up
python's min()/max()?)

Also, the `import matplotlib` should come before `import
matplotlib.pyplot`, as it would allow one to take certain actions such as
selecting backend prior to committing to it by importing pyplot.

As for documenting this, I still think there is some other problem going on
here that is complicating things. The difficulties you are encountering
doesn't seem to be normal and I wonder if something else causing problems.

**Comment by jenshnielsen** [comment]
IPYMPL is not documented because it's very early alpha state. Which should be indicated by the readme and the fact that it is V 0.0.2 but it will be better documented in the future once it is ready to use. 

The existing notebook backend within matplotlib has some bugs and limitations where this is probably one of them we may be able to fix this bug there in the mean time until ipympl is ready

That being said I cannot reproduce either of your issues with ipympl
https://gist.github.com/jenshnielsen/6af81090e8d243c3ea2926abef77b19f

**Comment by jenshnielsen** [comment]
The set_dpi warning is probably only happing when using this on retina and other hidp monitors and is fixed by #5383 for me. I could not reproduce it originally on a regular dpi monitor.

**Comment by jenshnielsen** [comment]
`%matplotlib ipympl` is added in https://github.com/ipython/ipython/pull/9885

**Comment by SylvainCorlay** [comment]
@michalsustr importing ipympl should switch the backend to ipympl.
